content/english/documentation/configuration.md
---
title: "Configuration"
description: ""
draft: false
weight: 2
bg_image: ""
toc : true
---

Many aspects of the ICU can be configured in the `config.json` file. 

## General Configuration

* `screen_width (int, float)` : window width.
* `screen_height (int, float)` : window height.
* `screen_size ` : window size. Default: (800, 700).
* `screen_x (int, float)` : window x position.
* `screen_y (int, float)` : window y position.
* `screen_position ` : window position. Default: (0, 0).
* `screen_full (bool)` : window full screen ?. Default: False.
* `screen_resizable (bool)` : window resizable ?. Default: True.
* `screen_aspect ` : window aspect ratio.
* `screen_min_size ` : minimum window size. Default: (100, 100).
* `screen_max_size ` : maximum window size. Default: (2000, 2000).
* `background_colour (str)` : cosmetic, window background colour. Default: grey.

----------------------------------------- 

## Task Configuration
Tasks can be enabled/disabled using the `task` configuration options:
* `system (bool)` : enable/disable tracking task.
* `track (bool)` : enable/disable system task.
* `fuel (bool)` : enable/disable fuel task.

The above configuration will disable both the tracking and fuel monitoring tasks, leaving the system monitoring task. All tasks are enabled by default. 

<hr>

## Task: Tracking {#tracking}

* `schedule ` : event schedule. Target drift, each event moves the target by `step` amount. Default: schedule: `[[100]]`.
* `step (int, float)` : distance (pixels) the Target moves on each event. Default: 2.
* `invert (bool)` : invert controls for tracking. Default: False.

### Example (1) - Target
```
"Target:0" : { "step" : 2 }
```

In the above configuration, the target will move 2 pixels per `move` event and according to its movement `schedule`.

<hr>

## Task: Fuel Monitoring {#fuel-monitoring}

The fuel monitoring task involves tanks and pumps, each have their own configuration options. Pumps move fuel between their associated tanks. 

### Tank

Tanks: FuelTank:A, FuelTank:B, FuelTank:C, FuelTank:D, FuelTank:E, FuelTank:F

* `burn_rate (int, float)` : rate at which fuel is burnt units/second.
* `accept_position (int, float)` : position of the acceptable level of fuel (for a main tank).
* `accept_proportion (int, float)` : proportion of acceptability of fuel (for a main tank), fuel levels within the range: position +- proportion are acceptable..
* `capacity (int, float)` : capacity of the tank.
* `fuel (int, float)` : initial fuel level.
* `fuel_colour (str)` : cosmetic, colour of the fuel.
* `background_colour (str)` : cosmetic, background colour of the tank.
* `outline_colour (str)` : cosmetic, outline colour of the tank.
* `outline_thickness (int)` : cosmetic, outline thickness of the tank.

### Pump

Pumps: Pump:AB, Pump:BA, Pump:FD, Pump:EA, Pump:CA, Pump:EC, Pump:DB, Pump:FB

* `schedule ` : event schedule. Each event causes the pump to fail/repair (repeating every two events). Default: schedule: `[[uniform(1000,20000)]]`.
* `flow_rate (int)` : transfer rate of fuel from one tank to another (units/second). Default: 100.
* `event_rate (int)` : number of events to trigger /second. Default: 10.
* `scale (float)` : cosmetic, the display scale of the pump.

### Example (1) - Tank Options
```
"FuelTank:A" : {"capacity":2000, "fuel":1000, "burn_rate":10}
```
In the above configuration, tank A will have a maximum capacity of 2000 units of fuel and start with 1000 units of fuel and burn 10 units of fuel per second. Note that only the main tanks (`FuelTank:A` and `FuelTank:B`) can burn fuel.

### Example (2) - Pump Options
```
"Pump:AB" : { "flow_rate": 10, "event_rate": 1}
```
For the above configuration for pump AB, 1 `transfer` event is triggered every second, each event transferring 10 units of fuel from tank A to tank B. Increasing the event rate to 10 would have the effect of 10 events being triggered (uniformly over a 1 second) with each event transferring 1 unit of fuel. Note that higher event rates lead to a smoother fuel animation but at the cost of many more events being generated.

<hr>

## Task: System Monitoring {#system-monitoring}

### Warning Light

Warning Lights: WarningLight:0, WarningLight:1

* `schedule ` : event schedule. Each event switches the light to the undesired state (off for WarningLight:0 and on for WarningLight:1). Default: schedule: `[[uniform(1000,10000)]]`.
* `grace (int, float)` : grace period (the time after which the light may turn on/off after a user has clicked). Default: 2.
* `key (str)` : key-binding shortcut, may be used instead of clicking. Default: 0/1.
* `state (int)` : initial state (on or off). Default: <K>.
* `on_colour (str)` : cosmetic, colour of the on state.
* `off_colour (str)` : cosmetic, colour of the off state.
* `outline_colour (str)` : cosmetic, colour of the outline.
* `outline_thickness (int)` : cosmetic, thickness of the outline.

### Example (1) - Warning Light Options
```
"WarningLight:0" : {"grace":1, "state":0}
```
In the above configuration, the warning light will start in state 0 (off) and will wait at least 1 second before turning back to an undesired state (according to its `schedule`).

### Scale

Scales: Scale:0, Scale:1, Scale:2, Scale:3, ... 

* `schedule ` : event schedule. Each events moves the slider up/down 1 place.. Default: schedule: `[[uniform(1000,10000)]]`.
* `key (str)` : key-binding shortcut, may be used instead of clicking. Default: <K>.
* `size (int)` : size of the scale (number of positions that the slider may be in). Default: 11.
* `position (int)` : initial position of the slider. Default: 5.
* `background_colour (str)` : cosmetic, background colour of the scale.
* `outline_colour (str)` : cosmetic, outline colour of the scale (and slider).
* `outline_thickness (int)` : cosmetic, outline thickness of the scale (and slider).
* `slider_colour (str)` : cosmetic, colour of the slider.


### Example (2) - Scale Options
```
"Scale:0" : {"size":11, "position":0},
```
In the above configuration, scale 0 will have 11 possible slider position and the slider will start at position 0 (at the top of the scale).

<hr>



## Input Configuration

Different types of input can be configured with the following options: 

* `mouse (bool)` : enable/disable mouse input. Default: True.
* `keyboard (bool)` : enable/disable keyboard input. Default: True.
* `joystick (bool)` : enable/disable joystick input. Default: False.

Input also contains a group of options for eye tracking.

### Eye Tracking Configuration

* `enabled (bool)` : enable eye tracking (requires a device to be connected).
* `calibrate (bool)` : calibrate the eye-tracker.
* `sample_rate (int)` : sample rate for the eye-tracker (if configurable on the device).
* `stub (bool)` : use the mouse as a stub for an eye tracking device, functions exactly as an eyetracker (useful for testing).

### Example (1) - Input

```
"input" : { "mouse" : true, 
            "eyetracker" : {
                "enabled" : false,
                "stub" : false }} 
```

The above configuration will enable the mouse as an input method, and completely disable the eye tracking. 

## Overlay Configuration

ICU provides support for an overlay, including highlighting widgets, displaying warnings etc. The following options are available under `overlay`.

* `enable (bool)` : enable/disable overlay (highlighting, arrows etc). Default: True.
* `arrow (bool)` : (IN PROGRESS) enable/disable arrows. Default: True.
* `transparent (bool)` : transparent highlights ?. Default: True.
* `outline (bool)` : outlined highlights ?. Default: True.
* `highlight_thickness (int)` : cosmetic, highlight outline thickness.
* `highlight_colour (str)` : cosmetic, highlight colour.
* `arrow_colour (colour)` : (IN PROGRESS) cosmetic, colour of the arrow
* `arrow_scale (float)` : (IN PROGRESS) cosmetic, scale of the arrow

----------------------------------------- 

## Event Configuration

Internal ICU events can be triggered according to schedules that are defined in the configuration. There are four kinds of schedules that can be specified, single, multi, repeating, and custom. Schedules are specified in the `schedule` option of each component. 

### Single-Event Schedule

A single event schedule is specified simply as an integer, `"Scale:0" : 1000` will cause an event to happen 1000ms after ICU begins.

```
t: 0 ------ 1000 ------ 2000 ------ 3000 ---
e: ---------- | ---------------------------- 
```

### Multi-Event Schedule

Multiple events can be schedules using a list of integer values, specified with square brackets `[ ]`, `"Scale:0" : [1000, 2000]` will trigger an event at 1000ms then wait 2000ms before triggering a second event.

```
t: 0 ------ 1000 ------ 2000 ------ 3000 ---
e: ---------- | --------------------- | ---- 
```

### Repeating-Event Schedule

Events schedules can be repeated by enclosing a list in a second pair of square brackets, `"Scale:0" : [[500, 1000]]`.

```
t: 0 ------- 1000 ------- 2000 ------- 3000 ---
e: ---- | ---------- | ---- | ---------- | ---- 
```

### Custom-Event Schedule

More complex schedules can be specified programmatically by subclassing `icu.config.Distribution`, subclasses will automatically be registered with the config module. Custom schedule are specified in `config.json` as `<name>(<arg1>, <arg2>, ...,<argn>)` where `<name>` is the class name and `<argi>` is an argument for the `__init__` method. Arguments should be parsed by the class. The method `sample(self) -> int` should be defined, which when called returns the time to wait before triggering the next event. 

Some common distributions have been defined for convenience.

#### Uniform Distribution
Takes the form `uniform(a,b)` where `a` and `b` are the lower and upper bounds (int > 0) of the domain, `uniform(0, 1000)`, will trigger an event after a random time in the interval 0-1 seconds.
#### Normal Distribution
Takes the form `normal(mu, sigma)` where `mu` and `sigma` are the mean and standard deviation of a Gaussian distribution, `normal(1000,100)`, will trigger an event at a random time centered around 1000ms. Note that negative sample values are treated as 0 (i.e. immediately trigger the event).

Custom-Event Schedules can be combined with multi or repeating schedules as follows: 

### Example (1) - Pump Scheduling

```
"Pump:AB" : { "schedule": [["uniform(0,10000)",1000]] }
```

In the above, the pump will fail randomly at a time between 0-10 seconds, then wait 1 second before being repaired, and repeat.

-------------------------------

## Cosmetic Options

Many of the display properties of widgets in ICU can be configured, common options include:
* `background_colour` : set the background colour of a widget.
* `outline_colour`    : set the outline colour of a widget. (default black)
* `outline_thickness` : set the outline thickness of a widget (default 3)

All colour options may be specified as a string e.g. `"red"` (see the [full list](http://www.science.smith.edu/dftwiki/index.php/Color_Charts_for_TKinter)) or as a hexadecimal string e.g. `"#ff0000"`.

A number of widgets have specific cosmetic options (see above).












icu/config.py
import json
import os
import re
import random
import copy

from types import SimpleNamespace
from collections import defaultdict

from itertools import cycle, islice


class ConfigurationError(Exception):
    pass

# size 800,800
# banner 25

# scales 240, 350

# fuel padding 250 


# =================================== # =================================== # =================================== # 
# =================================== # ======== ALL CONFIG OPTIONS ======= # =================================== # 
# =================================== # =================================== # =================================== #

def tasks():
    return SimpleNamespace(**{k:k for k in ['system', 'fuel', 'track']})

def default_config_screen():
    return dict(#screen_width=None,  #handled by post processing
                #screen_height=None, #handled by post processing
                screen_size=(800,700),
                #screen_x = None, #handled by post processing
                #screen_y = None, #handled by post processing
                screen_position=(0,0),
                screen_min_size=(100,100),
                screen_max_size=(2000,2000),
                screen_full=False,
                screen_resizable=True,
                screen_aspect=None,
                background_colour='grey',
                shutdown = -1)

def default_task_options():
    return   {"system" : True, "track" : True,"fuel" : True}

def default_event_schedule():
    return {"Scale:0" :         default_scale_schedule(),
            "Scale:1" :         default_scale_schedule(),
            "Scale:2" :         default_scale_schedule(),
            "Scale:3" :         default_scale_schedule(),
            "WarningLight:0" :  default_warning_light_schedule(),
            "WarningLight:1" :  default_warning_light_schedule(),
            "Target:0" :        default_target_schedule()}

def default_event_generator():
    return {"ScaleComponent" : "ScaleEventGenerator"}

def default_scale_schedule():
    return uniform(1000,10000) 

def default_warning_light_schedule():
    return uniform(1000,10000) 

def default_pump_schedule():
    return uniform(0,20000)

def default_target_schedule():
    return cycle([300])

# TODO config should contain everything --- create a default for all widgets!
def default_scales():
    return {"Scale:0" : default_scale("<F1>"),
            "Scale:1" : default_scale("<F2>"),
            "Scale:2" : default_scale("<F3>"),
            "Scale:3" : default_scale("<F4>")}

def default_scale(key):
    return {'size':11, 'position':5, 'key':key}

def default_warning_lights():
    return {'WarningLight:0' : {'state':1, 'key':"<F5>", 'grace':2}, 'WarningLight:1' : {'state':0, 'key':"<F6>", 'grace':2}}

def  default_tanks():
    return {
        "FuelTank:A" : {"capacity":2000, "fuel":1000, "burn_rate":6, "accept_position":0.5, "accept_proportion":0.3},
        "FuelTank:B" : {"capacity":2000, "fuel":1000, "burn_rate":6, "accept_position":0.5, "accept_proportion":0.3},
        "FuelTank:C" : {"capacity":1000, "fuel":100},
        "FuelTank:D" : {"capacity":1000, "fuel":100},
        "FuelTank:E" : {"capacity":1000, "fuel":1000},
        "FuelTank:F" : {"capacity":1000, "fuel":1000}}

def default_pumps():
    default_pump = { "flow_rate": 100, "event_rate": 10, "state": 1}

    return {"Pump:AB" : copy.deepcopy(default_pump),
            "Pump:BA" : copy.deepcopy(default_pump), 
            "Pump:FD" : copy.deepcopy(default_pump),
            "Pump:EA" : copy.deepcopy(default_pump),
            "Pump:CA" : copy.deepcopy(default_pump),
            "Pump:EC" : copy.deepcopy(default_pump),
            "Pump:DB" : copy.deepcopy(default_pump),
            "Pump:FB" : copy.deepcopy(default_pump)}
    

def default_tracking():
    return {} #TODO

def default_input():
    return {"mouse" : True,
            "keyboard" : True,
            "joystick" : False,
            "eyetracker" : {
                "stub" : True,
                "enabled" : True,
                "sample_rate" : 100,
                "calibrate" : False
            }}


def default_overlay():
    return dict(enable=True, transparent=True, outline=True, arrow=True)

def default_config():
    return dict(**default_config_screen(), task=default_task_options(), schedule=default_event_schedule(), overlay=default_overlay(), input=default_input(),
                **default_scales(), **default_warning_lights(), **default_tanks(), **default_pumps(), **default_tracking())

ScreenOptions = SimpleNamespace(**{k:k for k in default_config_screen().keys()})
# TODO other options

# =================================== # =================================== # =================================== # 

# =================================== # =================================== # =================================== # 
# =================================== # ========= DISTRIBUTIONS =========== # =================================== # 
# =================================== # =================================== # =================================== #

class Distribution: #must be an iterable...
    
    def __iter__(self):
        return self

    def __next__(self):
        return self.sample()

class uniform(Distribution):

    def __init__(self, a, b):
        try:
            #assert a >= 0 and b >= 0
            a = float(a)
            b = float(b)
            self.a = min(a,b)
            self.b = max(a,b)
        except:
            raise ConfigurationError("Invalid arguments for uniform distribution: {0}, {1}, must be numbers > 0".format(a,b))

    def sample(self):
        return random.uniform(self.a, self.b)

class normal(Distribution):

    def __init__(self, mu, sigma, decay=1.):
        try:
            self.mu = float(mu)
            self.sigma = float(sigma)
            self.decay = float(decay) #multiplicative decay?
        except:
            raise ConfigurationError("Invalid arguments for uniform distribution: {0}, {1}, must be numbers".format(mu, sigma))

    def sample(self):
        self.mu = self.mu * self.decay
        return max(0, random.gauss(self.mu, self.sigma)) #not thread safe?

distributions = lambda: {k.__name__:k for k in Distribution.__subclasses__()}

# =================================== # =================================== # =================================== # 

def is_type(*types): #validate type(s) 
    def _is_type(**kwargs): #expects a singleton dictionary
        assert len(kwargs) == 1
        k = next(iter(kwargs.keys()))
        v = kwargs[k]
        #print(k,v,types)
        if not isinstance(v, tuple(types)):
            raise ConfigurationError("Invalid value '{0}' for '{1}', must be one of type: {2}.".format(v, k, tuple([t.__name__ for t in types])))
        return v
    return _is_type

def is_coord():
    def _is_coord(**kwargs):
        k = next(iter(kwargs.keys()))
        v = kwargs[k]
        if isinstance(v, list) and len(v) == 2 and isinstance(v[0], (int, float)) and isinstance(v[1], (int,float)):
            return tuple(v)
        else:
            raise ConfigurationError("Invalid value '{0}' for '{1}', must be a list of length 2 and contain only numbers.".format(v, k))
    return _is_coord

def condition(cond):
    def _condition(**kwargs):
        assert len(kwargs) == 1
        k = next(iter(kwargs.keys()))
        v = kwargs[k]
        if not cond(v):
            raise ConfigurationError("Invalid value '{0}' for '{1}'.".format(v, k))
        return v
    return _condition
        
def get_option(k, group, _options=None):
    if _options is None:
        _options = options

    k = k.split(":")[0]
    if k in _options:
        option = _options[k]
        assert isinstance(option, Option)
        if option.group == group:
            return option

    raise ConfigurationError("Invalid config option: \"{0}\" for config group: \"{1}\"".format(k, group))

def validate_options(group, _options=None):
    return (lambda **kwargs : {k:get_option(k, group, _options=_options)(**{k:v}) for k,v in next(iter(kwargs.values())).items()})

class Option:
    
    def __init__(self, group, fun):
        self.group =  group
        self.__fun = fun

    def __call__(self, **kwargs):
        return self.__fun(**kwargs)

# CONFIG OPTIONS ARE ALL DEFINED IN THE DICTIONARY BELOW - EACH SHOULD BE VALIDATED



target_options = dict(
    step               = Option('target', is_type(int, float)),
    invert             = Option('target', is_type(bool))
)

warninglight_options = dict(
    grace               = Option('warning_light', is_type(int, float)),
    key                 = Option('warning_light', is_type(str)),
    state               = Option('warning_light', is_type(int)),
    on_colour           = Option('warning_light', is_type(str)),
    off_colour          = Option('warning_light', is_type(str)),
    outline_colour      = Option('warning_light', is_type(str)),
    outline_thickness   = Option('warning_light', is_type(int))
)

scale_options = dict(
    key                 = Option('warning_light', is_type(str)),
    size                = Option('scale', is_type(int)),
    position            = Option('scale', is_type(int)),
    background_colour   = Option('scale', is_type(str)),
    outline_colour      = Option('scale', is_type(str)),
    outline_thickness   = Option('scale', is_type(int)),
    slider_colour       = Option('scale', is_type(str))
)

pump_options = dict(
    flow_rate           = Option('pump', is_type(int)),
    event_rate          = Option('pump', is_type(int)),
    #cosmetic
    scale               = Option('pump', is_type(float))
)

tank_options = dict(
    burn_rate           = Option('tank', is_type(int, float)),
    accept_position     = Option('tank', is_type(int, float)),
    accept_proportion   = Option('tank', is_type(int, float)),
    capacity            = Option('tank', is_type(int, float)),
    fuel                = Option('tank', is_type(int, float)),
    # cosmetic
    fuel_colour         = Option('tank', is_type(str)),
    background_colour   = Option('tank', is_type(str)),
    outline_colour      = Option('tank', is_type(str)),
    outline_thickness   = Option('tank', is_type(int)),
)

eyetracker_options = dict(
    enabled             = Option('eyetracker', is_type(bool)),
    stub                = Option('eyetracker', is_type(bool)),
    calibrate           = Option('eyetracker', is_type(bool)),
    sample_rate         = Option('eyetracker', is_type(int)),
)

options = dict(

            main            = Option('-', validate_options('main')),

            #TODO validate these arguments?
            Pump            = Option('main', validate_options('pump', _options=pump_options)),
            Target          = Option('main', validate_options('target', _options=target_options)),
            WarningLight    = Option('main', validate_options('warning_light', _options=warninglight_options)),
            Scale           = Option('main', validate_options('scale', _options=scale_options)),
            FuelTank        = Option('main', validate_options('tank', _options=tank_options)),

            screen_width    = Option('main', is_type(int, float)),
            screen_height   = Option('main', is_type(int, float)),
            screen_size     = Option('main', is_coord()),
            screen_x        = Option('main', is_type(int, float)),
            screen_y        = Option('main', is_type(int, float)),
            screen_position        = Option('main', is_coord()),

            screen_full            = Option('main', is_type(bool)),
            screen_resizable       = Option('main', is_type(bool)),
            screen_aspect          = Option('main', is_coord()),
            screen_min_size        = Option('main', is_coord()),
            screen_max_size        = Option('main', is_coord()),
            background_colour       = Option('main', is_type(str)),
            
            schedule        = Option('main', lambda **kwargs: {k:Validator.is_schedule(**{k:v}) for k,v in next(iter(kwargs.values())).items()}),
            
            task            = Option('main', validate_options('task')),
            system          = Option('task', is_type(bool)),
            track           = Option('task', is_type(bool)),
            fuel            = Option('task', is_type(bool)),

            input           = Option('main',  validate_options('input')),
            mouse           = Option('input', is_type(bool)),
            keyboard        = Option('input', is_type(bool)),
            joystick        = Option('input', is_type(bool)),

            eyetracker      = Option('input', validate_options('eyetracker', _options=eyetracker_options)),

            generators      = Option('main', lambda **kwargs: None), #TODO
            shutdown          = Option('main', is_type(int, float)), # stop the system

            overlay             = Option('main',    validate_options('overlay')),
            enable              = Option('overlay', is_type(bool)),
            arrow               = Option('overlay', is_type(bool)),
            transparent         = Option('overlay', is_type(bool)),
            outline             = Option('overlay', is_type(bool)),
            highlight_thickness = Option('overlay', is_type(int)),
            highlight_colour    = Option('overlay', is_type(str)),
        
    )


import collections.abc

def update(d, u): #recursive dictionary update
    for k, v in u.items():
        if isinstance(v, collections.abc.Mapping):
            d[k] = update(d.get(k, {}), v)
        else:
            d[k] = v
    return d

class Validator:

    def validate_iter(k, v):
        for i in v:
            if not isinstance(i, (int, float)):
                raise ConfigurationError("Invalid value '{0}' for '{1}' in repeating schedule {2}, must be a number.".format(i, k, v))
        return v

    def validate_list(k, v):
        if len(v) > 0 and isinstance(v[0], list):
            if len(v) > 1:
                raise ConfigurationError("Invalid value '{0}' for '{1}', a repeating schedule is specified through the use of double square brackets: [[...]].".format(k, v))
            v = Validator.validate_iter(k, v[0])
            return cycle(v)
        else:
            v = Validator.validate_iter(k, v)
            return iter(v)

    def validate_str(k, v):
        pattern = '\w+\(((\w|\d)+,)*((\w|\d)+)?\)'
        r = re.match(pattern, re.sub(r"\s+", "", v))
        if r is not None:
            name, args = v.split('(')
            dists = distributions()
            if name in dists:
                args = args[:-1].split(",")
                try:
                    return dists[name](*args)
                except Exception as e:
                    raise ConfigurationError("Invalid value '{0}' for '{1}', failed to build distribution, perhaps the arguments were invalid.".format(v,k)) from e
            else:
                raise ConfigurationError("Invalid value '{0}' for '{1}', distribution not found, valid distributions include: {2}".format(v,k,tuple(distributions().keys())))


    def is_schedule(**kwargs): #validate schedule (number, list, tuple, str)
        k = next(iter(kwargs.keys()))
        v = kwargs[k]
        if isinstance(v, (int, float)): #schedule a single event
            return v
        elif isinstance(v, list):
            return Validator.validate_list(k, v)
        elif isinstance(v, str): #build schedule object
            return Validator.validate_str(k, v)
        raise ConfigurationError("Invalid value '{0}' for '{1}', must be a number, tuple, list or distribution.".format(v,k))


    def __call__(self, **kwargs):
        config =  validate_options('main')(**{'main':kwargs}) #{k:get_option(k, 'main')(**{k:v}) for k,v in kwargs.items()}

        # ====== POST PROCESSING ====== # 
        #TOOD move this somewhere more suitable
        result = default_config()
        result = update(result, config)
        #from pprint import pprint
        #pprint(result)

        result['screen_size'] = (result.get('screen_width', result['screen_size'][0]), result.get('screen_height', result['screen_size'][1]))
        result['screen_position'] = (result.get('screen_x', result['screen_position'][0]), result.get('screen_y', result['screen_position'][1]))
        result['screen_width'], result['screen_height'] = result['screen_size']
        result['screen_x'], result['screen_y'] = result['screen_position']
        return result

validate = Validator()

def save(path, **kwargs):
    """ Save config to a file.

    Args:
        path (str): path to config file (config file is always called 'config.json')
    """
    if not path.endswith('config.json'):
        path = os.path.join(path, 'config.json')
    if not os.path.exists(path):
        raise FileNotFoundError("Could not find config file at location: {0}".format(path))
    with open(path, 'w') as f:
        # maybe validate here?
        json.dump(kwargs, f, indent=4, sort_keys=True)
    
def load(path):
    """ Load a config file.

    Args:
        path (str): path to config file

    Returns:
        dict: a dictionary containing loaded config
    """
    path = os.path.abspath(path)
    if not path.endswith('config.json'):
        path = os.path.join(path, 'config.json')
    if not os.path.exists(path):
        raise FileNotFoundError("Could not find config file at location: {0}".format(path))
    print("USING CONFIG FILE: {0}".format(path))
    with open(path, 'r') as f: 
        data = json.load(f)
        result = validate(**data)
        return result

def reset(path):
    """ Reset config file to default values.

    Args:
        path (str): path to config file
    """
    if not path.endswith('config.json'):
        path = os.path.join(path, 'config.json')
    if not os.path.exists(path):
        raise FileNotFoundError("Could not find config file at location: {0}".format(path))
    with open(path, 'w') as f:
        json.dump(kwargs, f, indent=4, sort_keys=True)

if __name__ == "__main__":

    def run():
        path = 'icu/'
        save(path, **default_config())
        config = load(path)

        for k,v in config.items():
            print(k, v, type(v))
    run()

    #regex tests...
    from pprint import pprint
    for i in default_config_screen():
        print(i, ":")
    print()
    for i in default_event_schedule():
        print(i, ":")
    
    """
    def match(v):
        pattern = '\w+\(((\w|\d),)*((\w|\d)+)?\)'
        r = re.match(pattern, v)
        return r

    print(match('uniform(0,1000)'))
    print(match('uniform(10)'))
    print(match('uniform()'))
    """